src/compiler/crystal/semantic/semantic_visitor.cr

```diff
@@ -25,6 +25,24 @@ abstract class Crystal::SemanticVisitor < Crystal::Visitor
     @in_is_a = false
   end
 
+  def backup
+    old_current_type = @current_type
+    old_scope = @scope
+    old_exp_nest = @exp_nest
+    old_in_lib = @in_lib
+    old_in_c_struct_or_union = @in_c_struct_or_union
+    old_in_is_a = @in_is_a
+
+    yield
+
+    @current_type = old_current_type
+    @scope = old_scope
+    @exp_nest = old_exp_nest
+    @in_lib = old_in_lib
+    @in_c_struct_or_union = old_in_c_struct_or_union
+    @in_is_a = old_in_is_a
```

Comment:
Shouldn't this be guarded with `ensure` block?

```suggestion
    begin
      yield
    ensure
      @current_type = old_current_type
      @scope = old_scope
      @exp_nest = old_exp_nest
      @in_lib = old_in_lib
      @in_c_struct_or_union = old_in_c_struct_or_union
      @in_is_a = old_in_is_a
```

Reply:
Oh, great point. I think that's why what I was trying to achieve wasn't working (this enables restoring the current type and scope information in case of an error, but I didn't spend too much time on this because back then it was a nice to have)